### PR TITLE
Add VAE example

### PR DESCRIPTION
**Pull request by ywatanabex**
I have implemented VAE example, which has been suggested in #806

### CODE REVIEW

File: examples/vae/README.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Variational AutoEncoder
+
+This is a sample implementation of variational autoencoder.
+
+This metod is proposed by Kingma and Welling, Auto-Encoding Variational Bayes, 2014.
```
Comment (delta2323) [inline examples/vae/README.md]: method

File: examples/vae/README.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+
+This example requires scikit-learn to load the MNIST dataset.
```
Comment (delta2323) [inline examples/vae/README.md]: This example does not depend on scikit-learn.

File: examples/vae/net.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        if sigmoid:
+            return F.sigmoid(h2)
+        else:
+            return h2
+
+    def get_loss_func(self, C=1.0, L=1, train=True):
+        """Get loss function of VAE.
+
+        Args:
+            C (int): 1.0
```
Comment (delta2323) [inline examples/vae/net.py]: Please add the description of `C`.

File: examples/vae/net.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            L (int): Number of Monte Carlo samples used in encoded vector.
+            train (bool): if true loss_function is used for training.
+        """
+        def lf(x):
+            _, n_in = x.data.shape
+            mu, ln_var = self.encode(x)
+            batchsize, _ = mu.data.shape
```
Comment (delta2323) [inline examples/vae/net.py]: `batchsize = len(mu.data)` is better.

File: examples/vae/net.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    def get_loss_func(self, C=1.0, L=1, train=True):
+        """Get loss function of VAE.
+
+        Args:
+            C (int): 1.0
+            L (int): Number of Monte Carlo samples used in encoded vector.
+            train (bool): if true loss_function is used for training.
+        """
+        def lf(x):
+            _, n_in = x.data.shape
```
Comment (delta2323) [inline examples/vae/net.py]: `n_in = x.data.shape[1]`is better.

File: examples/vae/net.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+            # binarize input
+            xi = chainer.variable.Variable(
+                (x.data > self.xp.random.random(x.data.shape)).
+                astype(self.xp.int32), volatile=not train)
+            # reconstruction loss
+            rec_loss = 0
+            for l in range(L):
+                e = self.xp.random.normal(size=(batchsize, self.n_latent))
+                z = mu + F.exp(0.5 * ln_var) * e
+                rec_loss += (1.0/L) * \
```
Comment (delta2323) [inline examples/vae/net.py]: `1.0 / L` (insert spaces between an operator).

File: examples/vae/net.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            batchsize, _ = mu.data.shape
+            # binarize input
+            xi = chainer.variable.Variable(
+                (x.data > self.xp.random.random(x.data.shape)).
+                astype(self.xp.int32), volatile=not train)
+            # reconstruction loss
+            rec_loss = 0
+            for l in range(L):
+                e = self.xp.random.normal(size=(batchsize, self.n_latent))
+                z = mu + F.exp(0.5 * ln_var) * e
```
Comment (delta2323) [inline examples/vae/net.py]: We have a `Gaussian` Function for sampling from specified gaussian distribution. How about using it?
https://github.com/pfnet/chainer/blob/master/chainer/functions/noise/gaussian.py

File: examples/vae/net.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+                rec_loss += (1.0/L) * \
+                    sigmoid_cross_entropy(self.decode(z, sigmoid=False), xi)
```
Comment (delta2323) [inline examples/vae/net.py]: How about using `bernoulli_nll`? https://github.com/pfnet/chainer/blob/master/chainer/functions/loss/vae.py#L43
We do not have to binarize `x` in this case.

**Comment by ywatanabex** [comment]
Thank you for reviewing my code. I have updated the branch to follow your comments.

File: examples/vae/net.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+            return h2
+
+    def get_loss_func(self, C=1.0, L=1, train=True):
+        """Get loss function of VAE.
+
+        The loss value is equal to ELBO (Evidence Lower Bound)
+        multiplied by -1.
+
+        Args:
+            C (int): Usually this is 1.0. Can be changed to controll the
```
Comment (delta2323) [inline examples/vae/net.py]: control

File: examples/vae/train_vae.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    print('train mean loss={}, mean reconstruction loss={}'
+          .format(sum_loss / N, sum_rec_loss / N))
+
+    # evaluation
+    sum_loss = 0
+    sum_rec_loss = 0
+    for i in six.moves.range(0, N_test, batchsize):
+        x = chainer.Variable(xp.asarray(x_test[i:i + batchsize]),
+                             volatile='on')
+        loss_func = model.get_loss_func(L=10, train=False)
```
Comment (delta2323) [inline examples/vae/train_vae.py]: Can we change to control `L` as a parameter?

File: examples/vae/net.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+            L (int): Number of Monte Carlo samples used in encoded vector.
+            train (bool): if true loss_function is used for training.
+        """
+        def lf(x):
+            n_in = x.data.shape[1]
+            mu, ln_var = self.encode(x)
+            batchsize = len(mu.data)
+            # reconstruction loss
+            rec_loss = 0
+            for l in range(L):
```
Comment (delta2323) [inline examples/vae/net.py]: Use `six.moves.range` instead of `range`.

File: examples/vae/net.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+                z = F.gaussian(mu, ln_var)
+                rec_loss += F.bernoulli_nll(x, self.decode(z, sigmoid=False)) / \
+                    (L * batchsize * n_in)
```
Comment (delta2323) [inline examples/vae/net.py]: Division by `n_in` and multiplication by `n_in` in the next line are squashed.

File: examples/vae/train_vae.py [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+    print('epoch', epoch)
+
+    # training
+    perm = np.random.permutation(N)
+    sum_loss = 0       # total loss
+    sum_rec_loss = 0   # reconstruction loss
+    for i in six.moves.range(0, N, batchsize):
+        x = chainer.Variable(xp.asarray(x_train[perm[i:i + batchsize]]))
+
+        # Pass the loss function (Classifier defines it) and its arguments
```
Comment (delta2323) [inline examples/vae/train_vae.py]: Remove remove this comment (this comment is not appropriate in this example).

File: examples/vae/train_vae.py [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+
+# show results
+if args.gpu >= 0:
+    model.to_cpu()
+
+
+# original images and reconstructed images
+def save_images(x, filename):
+    fig, ax = plt.subplots(3, 3, figsize=(9, 9), dpi=100)
+    for (ai, xi) in zip(ax.flatten(), x):
```
Comment (delta2323) [inline examples/vae/train_vae.py]: It is just a matter of taste but we can remove parentheses from `(ai, xi)`.

**Comment by delta2323** [comment]
Besides reconstruction of input images, how about generating images by sampling `z` from uniform distribution and decoding it by trained decoder?

File: examples/vae/net.py [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+
+    def get_loss_func(self, C=1.0, L=1, train=True):
+        """Get loss function of VAE.
+
+        The loss value is equal to ELBO (Evidence Lower Bound)
+        multiplied by -1.
+
+        Args:
+            C (int): Usually this is 1.0. Can be changed to controll the
+                second term of ELBO bound, which works as regularization.
```
Comment (delta2323) [inline examples/vae/net.py]: **the** second term

File: examples/vae/net.py [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+            L (int): Number of Monte Carlo samples used in encoded vector.
+            train (bool): if true loss_function is used for training.
```
Comment (delta2323) [inline examples/vae/net.py]: Capitalize `if`.

**Comment by delta2323** [comment]
I added comments again. So could you check them?

**Comment by ywatanabex** [comment]
I have updated the code. Please review.

I add a code to generate images from z.
z is sampled from normal distribution because the prior of the latent space is the normal distribution.

**Comment by delta2323** [comment]
Thank you! I will review again.

File: examples/vae/net.py [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+        super(VAE, self).__init__(
+            # encoder
+            le1=L.Linear(n_in, n_h),
+            le2_mu=L.Linear(n_h, n_latent),
+            le2_ln_var=L.Linear(n_h, n_latent),
+            # decoder
+            ld1=L.Linear(n_latent, n_h),
+            ld2=L.Linear(n_h, n_in),
+        )
+        self.n_latent = n_latent
```
Comment (delta2323) [inline examples/vae/net.py]: This attribute is not used.

File: examples/vae/train_vae.py [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+    perm = np.random.permutation(N)
+    sum_loss = 0       # total loss
+    sum_rec_loss = 0   # reconstruction loss
+    for i in six.moves.range(0, N, batchsize):
+        x = chainer.Variable(xp.asarray(x_train[perm[i:i + batchsize]]))
+        optimizer.update(model.get_loss_func(), x)
+        if epoch == 1 and i == 0:
+            with open('graph.dot', 'w') as o:
+                g = computational_graph.build_computational_graph(
+                    (model.loss, ), remove_split=True)
```
Comment (delta2323) [inline examples/vae/train_vae.py]: `remove_split` option is left just for backward compatibility. So it is better to remove it.

File: examples/vae/train_vae.py [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+# Save the model and the optimizer
+print('save the model')
+serializers.save_npz('mlp.model', model)
+print('save the optimizer')
+serializers.save_npz('mlp.state', optimizer)
+
+
+# show results
+if args.gpu >= 0:
+    model.to_cpu()
```
Comment (delta2323) [inline examples/vae/train_vae.py]: `to_cpu` does nothing when `Link` is on CPU. So we do not have to switch based on `args.cpu`.

File: examples/vae/train_vae.py [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+    fig, ax = plt.subplots(3, 3, figsize=(9, 9), dpi=100)
+    for ai, xi in zip(ax.flatten(), x):
+        ai.imshow(xi.reshape(28, 28))
+    fig.savefig(filename)
+
+train_ind = [1, 3, 5, 10, 2, 0, 13, 15, 17]
+x = chainer.Variable(np.asarray(x_train[train_ind]), volatile='on')
+x1 = model(x)
+save_images(x.data, 'train')
+save_images(x1.data, 'train_ae')
```
Comment (delta2323) [inline examples/vae/train_vae.py]: I think `train_reconstructed` is more preferable as the output filename, although it is the matter of tastes.

File: examples/vae/train_vae.py [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+
+test_ind = [3, 2, 1, 18, 4, 8, 11, 17, 61]
+x = chainer.Variable(np.asarray(x_test[test_ind]), volatile='on')
+x1 = model(x)
+save_images(x.data, 'test')
+save_images(x1.data, 'test_ae')
```
Comment (delta2323) [inline examples/vae/train_vae.py]: ditto (`test_reconstructed`)

**Comment by delta2323** [comment]
I updated comments.

**Comment by delta2323** [comment]
LGTM :+1:

**Comment by delta2323** [comment]
Thank you for your contribution!
